File: lua/stack/03_medium_GenerateParentheses.lua
-- You are given an integer n. Return all well-formed parentheses strings that you can generate with n pairs of parentheses.

-- Example 1:

-- Input: n = 1

-- Output: ["()"]
-- Example 2:

-- Input: n = 3

-- Output: ["((()))","(()())","(())()","()(())","()()()"]

local function isEmpty(table)
    return next(table) == nil
end


local function generate_parenthesis(n)
    local result = {}
    local stack = {}
    table.insert(stack, {'', 0, 0})
    while not isEmpty(stack) do
        local current = table.remove(stack)
        local str = current[1]
        local open = current[2]
        local closed = current[3]
        if open == n and closed == n then
            table.insert(result, str)
        end
        if open < n then
            table.insert(stack, {str .. '(', open + 1, closed })
        end
        if closed < open then
            table.insert(stack, {str .. ')', open, closed + 1 })
        end
    end
    return result
end


print("test n = 1")
local n = 1
local parentheses_combinations = generate_parenthesis(n)
for i, combination in ipairs(parentheses_combinations) do
    print(combination)
end


print("test n = 3")
local n = 3
local parentheses_combinations = generate_parenthesis(n)
for i, combination in ipairs(parentheses_combinations) do
    print(combination)
end

File: rust/stack/03_medium_GenerateParentheses.rs
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

// Example 1:

// Input: n = 3
// Output: ["((()))","(()())","(())()","()(())","()()()"]
// Example 2:

// Input: n = 1
// Output: ["()"]


// Constraints:

// 1 <= n <= 8



pub fn generate_parenthesis(n: i32) -> Vec<String> {
    let mut result:Vec<String> = Vec::new();
    let mut stack:Vec<(String, i32, i32)> = Vec::new();
    stack.push((String::from(""), 0, 0));
    while !stack.is_empty() {
        let (str, open, closed) = stack.pop().unwrap();
        if open == n && closed == n {
            result.push(str.clone());
        }
        if open < n {
            let new_str = str.clone() + "(";
            stack.push((new_str, open + 1, closed));
        }
        if closed < open {
            let new_str = str.clone() + ")";
            stack.push((new_str, open, closed + 1));
        }
    }
    result
}


pub fn main() {

    println!("{:?}", generate_parenthesis(1));
    println!("{:?}", generate_parenthesis(3));
}
